### DISCUSSION TITLE
Syntax Highlighter example renders a div inside pre

### DISCUSSION THREAD

**Discussion by jahirfiquitiva** [remarkjs/remark#684] [Q&A] [answered]
Hey

I'm trying to implement the example in the readme (https://github.com/remarkjs/react-markdown#use-custom-components-syntax-highlight)

But when the site is built, I end up with something like
```html
<pre>
    <div>
        <code>
            ...
        </code>
    </div>
<pre>
```

What can I do to make it render like this?
```html
<pre>
    <code>
        ...
    </code>
<pre>
```

Thanks in advance

**Comment by wooorm** [comment, marked as answer]
It doesn’t show that from what I can see: https://codesandbox.io/s/react-markdown-debug-forked-z54vv?file=/src/App.js.

**Reply by jahirfiquitiva** [reply]
Well, that sandbox doesn't look the same as this example in the readme
https://github.com/remarkjs/react-markdown#use-custom-components-syntax-highlight

**Reply by wooorm** [reply]
yeah I changed something unrelated. Regardless, it sounds like you have other code. Can you make your problem reproducible?

**Reply by jahirfiquitiva** [reply]
Here you go

https://codesandbox.io/s/react-markdown-debug-forked-t34pe?file=/src/App.js

You can inspect the rendered html and you'll see it is 
```html
<pre>
    <div>
        <code>
            ...
        </code>
    </div>
<pre>
```

**Reply by ChristianMurphy** [reply]
So the double `pre` issue doesn't appear.
Is the `div` between unexpected? It seems to make sense give there is a frame around the code sample.
This looks correct. :thinking:

**Reply by wooorm** [reply]
Wait, you said “Syntax Highlighter example renders a div inside pre and then another pre”.
But your issue describes something else? There are no pres in pres?

The div isn’t a problem.

If you really dislike it, maybe ask `syntax-highlighter` folks to add a `PreTag: false` option. Or PR it?
Alternatively:

```js
  pre({ node, ...props }) {
    return <>{props.children}</>
  },
  code({ node, className, ...props }) {
    const match = /language-(\w+)/.exec(className || "");
    return match ? (
      <SyntaxHighlighter
        language={match[1]}
        style={dark}
        {...props}
      />
    ) : (
      <code className={className} {...props} />
    );
  }
```

^-- this works

**Reply by jahirfiquitiva** [reply]
woops, I worded it badly, ... but yeah, the last code sample works perfectly! thanks @wooorm

**Reply by Anthony-Gaudino** [reply]
Above solution removes the `pre` tag that is inserted by `remark`.
Another solution is to remove the `pre` tag that is required by `react-syntax-highlighter` by passing a component to the `react-syntax-highlighter` `PreTag` prop that renders an empty element with just the children:

```typescript
/**
 * Avoids an issue where two `pre` tags appear nested on rendered Markdown
 * because `react-markdown` adds a `pre` tag by default and
 * `react-syntax-highlighter` requires it's own `pre` tag to also be rendered.
 *
 * This `CustomPreTag` component is used to replace the
 * `react-syntax-highlighter` `pre` tag by an empty element, with just the
 * nested `code` element as children. This way only the `react-markdown`
 * `pre` tag is rendered.
 */
const CustomPreTag = ({ children }: { children: React.ReactNode }) => (
  <>{children}</>
);

// Then when rendering the Markdown:
<ReactMarkdown
  children={markdown}
  remarkPlugins={[remarkGfm]}
  components={{
    code({ node, inline, className, children, ...props }) {
      const match = /language-(\w+)/.exec(className || "");
      return !inline && match ? (
        <SyntaxHighlighter
          children={String(children).replace(/\n$/, "")}
          style={ghcolors}
          language={match[1]}
          PreTag={CustomPreTag} // Use the component here
          {...props}
        />
      ) : (
        <code className={className} {...props}>
          {children}
        </code>
      );
    }
  }}
/>
```

**Comment by SalahAdDin** [comment]
Well, it is not solved for me:
```
~~~ts
import "@capacitor-community/uxcam"
import { Plugins } from "@capacitor/core"
import { LogEventProperty, UserProperty } from "@capacitor-community/uxcam"
import { appEnv } from "../env"
~~~

```

Never gets syntax higlighted but ever shows a `code` tag, why? Where is className comming from?

**Reply by ChristianMurphy** [reply]
@SalahAdDin this seems unrelated to the original question?
There's a working example of rendering in the selected answer https://github.com/remarkjs/remark/discussions/684#discussioncomment-616412
And as was noted in the several other unrelated place you've asked, it's documented in the readme as well https://github.com/remarkjs/react-markdown#use-custom-components-syntax-highlight
If the issue persist, could you start a new question, and include a code sandbox with an example of what you are trying? https://codesandbox.io/s/react-markdown-debug-9n4eg can be used as a starting point.
